File: reference/binary-search/4-Median-of-Two-Sorted-Arrays.go
import (
\t\fmt\
\t\math\
)
func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
    if len(nums1) > len(nums2) {
        nums1, nums2 = nums2, nums1
    }

    m, n := len(nums1), len(nums2)
    low, high := 0, m

    for low <= high {
        partitionSmallArray := (low + high) / 2
        partitionLargeArray := (m + n + 1) / 2 - partitionSmallArray

        maxLeftSmall := math.Inf(-1)
\t\tif partitionSmallArray > 0 {
\t\t\tmaxLeftSmall = float64(nums1[partitionSmallArray-1])
\t\t}

\t\tminRightSmall := math.Inf(1)
\t\tif partitionSmallArray < m {
\t\t\tminRightSmall = float64(nums1[partitionSmallArray])
\t\t}

\t\tmaxLeftLarge := math.Inf(-1)
\t\tif partitionLargeArray > 0 {
\t\t\tmaxLeftLarge = float64(nums2[partitionLargeArray-1])
\t\t}

\t\tminRightLarge := math.Inf(1)
\t\tif partitionLargeArray < n {
\t\t\tminRightLarge = float64(nums2[partitionLargeArray])
\t\t}

        if maxLeftSmall <= minRightLarge && maxLeftLarge <= minRightSmall {
\t\t\t// If total number of elements is even
\t\t\tif (m+n)%2 == 0 {
\t\t\t\treturn (math.Max(maxLeftSmall, maxLeftLarge) + math.Min(minRightSmall, minRightLarge)) / 2
\t\t\t}
\t\t\t// If total number of elements is odd
\t\t\treturn math.Max(maxLeftSmall, maxLeftLarge)
\t\t} else if maxLeftSmall > minRightLarge {
\t\t\t// Move partition left in the smaller array
\t\t\thigh = partitionSmallArray - 1
\t\t} else {
\t\t\t// Move partition right in the smaller array
\t\t\tlow = partitionSmallArray + 1
\t\t}
    }

    return 0.0
}

File: reference/binary-search/4-Median-of-Two-Sorted-Arrays.py
class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        if len(nums1) > len(nums2):
            nums1, nums2 = nums2, nums1
        
        len1, len2 = len(nums1), len(nums2)
        low, high = 0, len1

        while low <= high:
            partitionSmallArray = (low + high) // 2
            partitionLargeArray = (len1 + len2 + 1) // 2 - partitionSmallArray

            maxLeftSmall = float('-inf') if partitionSmallArray == 0 else nums1[partitionSmallArray - 1]
            minRightSmall = float('inf') if partitionSmallArray == len1 else nums1[partitionSmallArray]

            maxLeftLarge = float('-inf') if partitionLargeArray == 0 else nums2[partitionLargeArray - 1]
            minRightLarge = float('inf') if partitionLargeArray == len2 else nums2[partitionLargeArray]

            if maxLeftSmall <= minRightLarge and maxLeftLarge <= minRightSmall:
                if (len1 + len2) % 2 == 0:
                    return (max(maxLeftSmall, maxLeftLarge) + min(minRightSmall, minRightLarge)) / 2
                else:
                    return max(maxLeftSmall, maxLeftLarge)
            elif maxLeftSmall > minRightLarge:
                high = partitionSmallArray - 1
            else:
                low = partitionSmallArray + 1

            
